zebra/dplane_fpm_nl.c

```diff
@@ -1031,9 +1031,39 @@ static int fpm_nl_finish(struct zebra_dplane_provider *prov, bool early)
 	struct fpm_nl_ctx *fnc;
 
 	fnc = dplane_provider_get_data(prov);
-	stream_free(fnc->ibuf);
-	stream_free(fnc->obuf);
-	close(fnc->socket);
+
+	/* This is the late call: stop the thread. */
+	if (early == false) {
+		/* Stop the running thread. */
+		frr_pthread_stop(fnc->fthread, NULL);
+
+		/* Free all allocated resources. */
+		pthread_mutex_destroy(&fnc->obuf_mutex);
+		pthread_mutex_destroy(&fnc->ctxqueue_mutex);
+		stream_free(fnc->ibuf);
+		stream_free(fnc->obuf);
+		free(gfnc);
+		gfnc = NULL;
+		return 0;
+	}
+
+	/* Early call: disable all events and close socket. */
+	THREAD_OFF(fnc->t_ribreset);
+	THREAD_OFF(fnc->t_ribwalk);
+	THREAD_OFF(fnc->t_rmacreset);
+	THREAD_OFF(fnc->t_rmacwalk);
+	if (fnc->t_read)
```

Comment:
You can't test these like this, it's a race condition, just call the async cancel directly

Reply:
Fixed!